include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,512 @@
+#ifndef RBTREE_HEADER_H
+#define RBTREE_HEADER_H
+
+/* Adapted from https://github.com/BartoszMilewski/Okasaki/tree/master/RBTree */
+/* Implementing algorithms from                                               *
+ * https://matt.might.net/papers/germane2014deletion.pdf                      */
+
+#include <cassert>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#define CONSTRUCT_MSG_AND_THROW(msg)                                           \
+  do {                                                                         \
+    std::string emsg = std::string("[") + std::string(__func__)                \
+                       + std::string("] ") + std::string(msg);                 \
+    throw std::invalid_argument(emsg);                                         \
+  } while (false)
+
+//---               Ordered map on top of a red-black tree.                ---//
+
+// 1. No red node has a red child.
+// 2. Every path from root to empty node contains the same
+// number of black nodes.
+
+template <class T, class V>
+class RBTree {
+  enum Color { R, B, BB };
+
+  static Color minusOneColor(Color c) {
+    switch (c) {
+    case B: return R;
+    case BB: return B;
+    default: CONSTRUCT_MSG_AND_THROW("Unexpected color");
+    }
+  }
+
+  struct Node {
+    Node(Color c)
+        : _c(c)
+        , _s(0) { }
+    Color _c;
+    size_t _s;
```

Comment:
Underscore-prefixed member variables aren't UB, but I tend to avoid them in favour of suffixed names as it's a little bit easier to avoid ending up with UB if you rename something.

Reply:
I changed to names that are followed by `_` instead.